### PR TITLE
[FIR] Some checkers for exposed visibility

### PR DESCRIPTION
**Pull request by rapturemain**
Support for:
EXPOSED_PARAMETER_TYPE
EXPOSED_RECEIVER_TYPE
EXPOSED_FUNCTION_RETURN_TYPE
EXPOSED_PROPERTY_TYPE
EXPOSED_TYPEALIAS_EXPANDED_TYPE

### CODE REVIEW

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExposedVisibilityChecker.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        when (declaration) {
+            is FirTypeAlias -> checkTypeAlias(declaration, context, reporter)
+            is FirProperty -> checkProperty(declaration, context, reporter)
+            is FirFunction<*> -> checkFunction(declaration, context, reporter)
+        }
+    }
+
+    private fun checkTypeAlias(declaration: FirTypeAlias, context: CheckerContext, reporter: DiagnosticReporter) {
+        val expandedType = declaration.expandedConeType
+        val typeAliasVisibility = declaration.firEffectiveVisibility(declaration.session)
```
Comment (mglukhikh) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExposedVisibilityChecker.kt:30]: Please use just `.effectiveVisibility` here and below. It works more correctly for local classes.

File: compiler/testData/diagnostics/testsWithStdLib/InaccessibleInternalClass.fir.kt [hunk 2]
```diff
@@ -10,6 +10,6 @@ package kotlin.sequences
 import p.*
 
 interface I {
-    val v1: FilteringSequence
-    val v2: IndexingSequence<String>
+    <!EXPOSED_PROPERTY_TYPE!>val v1: FilteringSequence<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/testsWithStdLib/InaccessibleInternalClass.fir.kt:13]: Visibility of `FilteringSequence` class is `Internal`, while visibility of v1 is `Public`

File: compiler/testData/diagnostics/tests/callableReference/function/localConstructorFromLocalClass.kt [hunk 3]
```diff
@@ -3,9 +3,9 @@ import kotlin.reflect.KFunction0
 
 fun main() {
     class A
-    
+
     class B {
-        val x = ::A
-        val f: KFunction0<A> = x
+        <!EXPOSED_PROPERTY_TYPE!>val x = ::A<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/tests/callableReference/function/localConstructorFromLocalClass.kt]: Fixed

File: compiler/testData/diagnostics/tests/callableReference/kt15439_completeCall.kt [hunk 4]
```diff
@@ -2,11 +2,11 @@
 // !DIAGNOSTICS: -UNUSED_VARIABLE, -UNUSED_VARIABLE
 
 fun test() {
-    data class Pair<F, S>(val first: F, val second: S)
+    <!EXPOSED_FUNCTION_RETURN_TYPE!>data class Pair<F, S>(val first: F, val second: S)<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/tests/callableReference/kt15439_completeCall.kt]: Fixed

File: compiler/testData/diagnostics/tests/typealias/typeAliasAsBareType.fir.kt [hunk 5]
```diff
@@ -29,7 +29,7 @@ fun testWrong4(x: List<Any>) = x is ReadableList
 
 fun <T> testLocal(x: Any) {
     class C
-    typealias CA = C
+    <!EXPOSED_TYPEALIAS_EXPANDED_TYPE!>typealias CA = C<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/tests/typealias/typeAliasAsBareType.fir.kt]: Fixed

File: compiler/testData/diagnostics/tests/inner/localThisSuper.fir.kt [hunk 6]
```diff
@@ -12,7 +12,7 @@ class Outer : Trait {
             val s = super@Outer.bar()
             
             inner class Inner {
-                val t = this@Local
+                <!EXPOSED_PROPERTY_TYPE!>val t = this@Local<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/tests/inner/localThisSuper.fir.kt]: Fixed

File: compiler/testData/diagnostics/tests/callableReference/function/localConstructorFromExtensionInLocalClass.kt [hunk 7]
```diff
@@ -5,13 +5,13 @@ import kotlin.reflect.KFunction0
...
     class A
-    
+
     class B {
         fun Int.foo() {
             val x = ::A
             checkSubtype<KFunction0<A>>(x)
         }
-        fun A.foo() {
+        fun <!EXPOSED_RECEIVER_TYPE!>A<!>.foo() {
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/testData/diagnostics/tests/callableReference/function/localConstructorFromExtensionInLocalClass.kt]: Fixed

File: compiler/fir/analysis-tests/testData/resolve/localClasses/implicitInLocalClasses.kt [hunk 8]
```diff
@@ -20,7 +20,7 @@ fun main() {
         fun foo(x: Int) = inner.w
         fun bar(y: Int) = this.hashCode() + y > 0
 
-        val inner = Inner()
+        <!EXPOSED_PROPERTY_TYPE!>val inner = Inner()<!>
```
[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/fir/analysis-tests/testData/resolve/localClasses/implicitInLocalClasses.kt]: Fixed

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExposedVisibilityChecker.kt (hunk 1 above)
Comment (rapturemain) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirExposedVisibilityChecker.kt:30]: Now using the combination of `.effectiveVisibility` and `.firEffectiveVisibility(session) `

File: compiler/fir/analysis-tests/testData/resolve/visibility/exposedFunctionReturnType.kt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class A {
+    private class InnerA {
+
+    }
+}
+
+abstract class B {
+    <!EXPOSED_FUNCTION_RETURN_TYPE!>fun foo(str: String): A.InnerA<!>
```
[the remark this answers is not in the document]
Comment (mglukhikh) [reply compiler/fir/analysis-tests/testData/resolve/visibility/exposedFunctionReturnType.kt:8]: This problem is still actual.

[the remark this answers is not in the document]
Comment (rapturemain) [reply compiler/fir/analysis-tests/testData/resolve/visibility/exposedFunctionReturnType.kt]: Fixed

**Comment by mglukhikh** [comment]
Merged manually. Congratulations! 👍

**Pull request closed without merging**